src/custom/state/orders/updater.ts

```diff
@@ -0,0 +1,176 @@
+import { Web3Provider } from '@ethersproject/providers'
+import { Log, Filter } from '@ethersproject/abstract-provider'
+import { useLastCheckedBlock } from './hooks'
+import { updateLastCheckedBlock } from './actions'
+// import { PartialOrdersMap } from './reducer'
+
+// example of event watching + decoding without contract
+const transferEventAbi = 'event Transfer(address indexed from, address indexed to, uint amount)'
+const ERC20Interface = new utils.Interface([transferEventAbi])
+
+const TransferEvent = ERC20Interface.getEvent('Transfer')
+
+const TransferEventTopics = ERC20Interface.encodeFilterTopics(TransferEvent, [])
+
+// const decodeTransferEvent = (transferEventLog: Log) => {
+//   return ERC20Interface.decodeEventLog(TransferEvent, transferEventLog.data, transferEventLog.topics)
+// }
+
+type RetryFilter = Filter & { fromBlock: number; toBlock: number }
+
+const constructGetLogsRetry = (provider: Web3Provider) => {
+  // tries fetching logs
+  // if breaks, retries on half the range
+  const getLogsRetry = async ({ fromBlock, toBlock, ...rest }: RetryFilter): Promise<Log[]> => {
+    try {
+      console.log(`EventUpdater::Getting logs fromBlock: ${fromBlock} toBlock ${toBlock}`)
+      const logs = await provider.getLogs({
+        fromBlock,
+        toBlock,
+        ...rest
+      })
+      // console.log('logs', logs)
+      return logs
+    } catch (error) {
+      console.error(`Error getting logs fromBlock: ${fromBlock} toBlock ${toBlock}`, error)
+
+      // expect -- RPC Error: query returned more than 10000 results --
+      // if a different error - rethrow
+      if (!error?.message?.includes('query returned more than')) throw error
+
+      const midBlock = Math.floor((toBlock + fromBlock) / 2)
+
+      const [beforeMidLogs, afterMidLogs] = await Promise.all([
+        getLogsRetry({
+          fromBlock,
+          toBlock: midBlock
+        }),
+        getLogsRetry({
+          fromBlock: midBlock + 1,
+          toBlock
+        })
+      ])
+
+      return beforeMidLogs.concat(afterMidLogs)
+    }
+  }
+
+  return getLogsRetry
+}
+
+export function EventUpdater(): null {
+  const { chainId, library } = useActiveWeb3React()
+  // console.log('EventUpdater::library', library)
+  // console.log('EventUpdater::chainId', chainId)
+
+  const lastBlockNumber = useBlockNumber()
+  const lastCheckedBlock = useLastCheckedBlock({ chainId })
+  console.log('EventUpdater::lastCheckedBlock', lastCheckedBlock)
+  console.log('EventUpdater::lastBlockNumber', lastBlockNumber)
+
+  const dispatch = useDispatch<AppDispatch>()
+
+  // show popup on confirm
+  // for displaying fulfilled orders
+  const addPopup = useAddPopup()
+
+  const getLogsRetry = useMemo(() => {
+    if (!library) return null
+    return constructGetLogsRetry(library)
+  }, [library])
+
+  useEffect(() => {
+    if (!chainId || !library || !getLogsRetry || !lastBlockNumber) return
+    ;(window as any).getLogsRetry = (fromBlock: number, toBlock: number) => {
```

Comment:
not opposed to leave some globals to easy test. But what do you think of doing a util `registerGlobal` so we are more explicit and we remove from the hook ugly casting

Reply:
sure, but later, as this waterfall will cause to many changes